pdns/iputils.hh

```diff
@@ -568,134 +632,372 @@ private:
-      memcpy(addr, key.getNetwork().sin6.sin6_addr.s6_addr, sizeof(addr));
-      std::bitset<64> addr_low(be64toh(addr[1]));
-      std::bitset<64> addr_high(be64toh(addr[0]));
-      int bits = 0;
-      while(bits < key.getBits()) {
-        uint8_t val;
-        // we use high address until we are
-        if (bits < 64) val = addr_high[63-bits];
-        // past 64 bits, and start using low address
-        else val = addr_low[127-bits];
-
-        // we turn left on 0 and right on 1
-        if (val)
-          node = node->make_right();
-        else
-          node = node->make_left();
-        bits++;
+      if (d_root->left)
+        is_left = false;
+    } else
+      throw NetmaskException("invalid address family");
+
+    // we turn left on 0 and right on 1
+    int bits = 0;
+    for(; node && bits < key.getBits(); bits++) {
+      bool vall = key.getBit(-1-bits);
+
+      if (bits >= node->d_bits) {
+        // the end of the current node is reached; continue with the next
+        if (vall) {
+          if (node->left || node->assigned)
+            is_left = false;
+          if (!node->right) {
+            // the right branch doesn't exist yet; attach our key here
+            node = node->make_right(key);
+            break;
+          }
+          node = node->right.get();
+        } else {
+          if (!node->left) {
+            // the left branch doesn't exist yet; attach our key here
+            node = node->make_left(key);
+            break;
+          }
+          node = node->left.get();
+        }
+        continue;
       }
-      // only create node if not yet assigned
-      if (!node->node6) {
-        node->node6 = unique_ptr<node_type>(new node_type());
-        _nodes.insert(node->node6.get());
+      if (bits >= node->node->first.getBits()) {
+        // the matching branch ends here, yet the key netmask has more bits; add a
+        // child node below the existing branch leaf.
+        if (vall) {
+          if (node->assigned)
+            is_left = false;
+          node = node->make_right(key);
+        } else {
+          node = node->make_left(key);
+        }
+        break;
+      }
+      bool valr = node->node->first.getBit(-1-bits);
+      if (vall != valr) {
+        if (vall)
+          is_left = false;
+        // the branch matches just upto this point, yet continues in a different
+        // direction; fork the branch.
+        node = node->fork(key, bits);
+        break;
       }
-      value = node->node6.get();
     }
-    // assign key
+
+    if (node->node->first.getBits() > key.getBits()) {
+      // key is a super-network of the matching node; split the branch and
+      // insert a node for the key above the matching node.
+      node = node->split(key, key.getBits());
+    }
+
+    if (node->left)
+      is_left = false;
+
+    node_type* value = node->node.get();
+
+    // overwrite netmask, since unmasked address may differ for an existing node
+    // (do we even care?)
```

Comment:
I did that now. However, the Recursor cache unit test started failing on one item. I fixed it by normalizing the netmask before storing it into the cache (in the replace() method). You should give that a look.

Reply:
Right, that's because the new NT stores and returns a normalized version of the EDNS Client Subnet scope Netmask, and we use that to lookup the best entry in the records cache. But the record cache itself stores the non-normalized version of the Netmask, so it fails, because `Netmask::operator<` compares the non-masked `ComboAddress` when everything else matches.
I think it would be nice to add the following change, in addition to the one you added:
```
diff --git pdns/recursor_cache.hh pdns/recursor_cache.hh
index 7c5557f25..00348c4b5 100644
--- pdns/recursor_cache.hh
+++ pdns/recursor_cache.hh
@@ -73,7 +73,7 @@ private:
   struct CacheEntry
   {
     CacheEntry(const boost::tuple<DNSName, uint16_t, Netmask>& key, bool auth):
-      d_qname(key.get<0>()), d_netmask(key.get<2>()), d_state(Indeterminate), d_ttd(0), d_qtype(key.get<1>()), d_auth(auth)
+      d_qname(key.get<0>()), d_netmask(key.get<2>().getNormalized()), d_state(Indeterminate), d_ttd(0), d_qtype(key.get<1>()), d_auth(auth)
     {
     }
```

But we should ponder if `Netmask::operator<` is currently doing the right thing, or it should compare the masked version of the `ComboAddress` instead. In fact, I'm not even sure why we keep the non-masked version at all?

Reply:
Applied your proposed change.

I did dare changing the masking behavior.

Reply:
Thanks! I'll open a separate issue for `Netmask::operator<`.